File: tests/src/org.joda/joda-convert/2.2.4/src/test/java/org_joda/joda_convert/AnnotationStringConverterFactoryTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_joda.joda_convert;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.joda.convert.FromString;
import org.joda.convert.FromStringFactory;
import org.joda.convert.StringConvert;
import org.joda.convert.ToString;
import org.junit.jupiter.api.Test;

public class AnnotationStringConverterFactoryTest {
    @Test
    public void findsAnnotatedMethodsConstructorsInterfacesAndFactories() {
        StringConvert convert = StringConvert.create();

        CharSequenceConstructorFixture fromConstructor =
                convert.convertFromString(CharSequenceConstructorFixture.class, "char-sequence");
        StringConstructorFixture stringConstructor =
                convert.convertFromString(StringConstructorFixture.class, "string");
        MethodAnnotatedValue fromMethod = convert.convertFromString(MethodAnnotatedValue.class, "method");
        InterfaceAnnotatedValue fromInterface =
                convert.convertFromString(InterfaceAnnotatedValue.class, "interface");
        FactoryAnnotatedValue fromFactory =
                convert.convertFromString(FactoryAnnotatedValue.class, "factory");

        assertEquals("char-sequence", convert.convertToString(fromConstructor));
        assertEquals("string", convert.convertToString(stringConstructor));
        assertEquals("method", convert.convertToString(fromMethod));
        assertEquals("interface", convert.convertToString(ConvertibleInterface.class, fromInterface));
        assertEquals("factory", convert.convertToString(fromFactory));
    }

    public static final class MethodAnnotatedValue {
        private final String value;

        MethodAnnotatedValue(String value) {
            this.value = value;
        }

        @ToString
        public String asText() {
            return value;
        }

        @FromString
        public static MethodAnnotatedValue parse(String value) {
            return new MethodAnnotatedValue(value);
        }
    }

    public interface ConvertibleInterface {
        @ToString
        String asText();

        @FromString
        static InterfaceAnnotatedValue parse(String value) {
            return new InterfaceAnnotatedValue(value);
        }
    }

    public static final class InterfaceAnnotatedValue implements ConvertibleInterface {
        private final String value;

        InterfaceAnnotatedValue(String value) {
            this.value = value;
        }

        @Override
        public String asText() {
            return value;
        }
    }

    @FromStringFactory(factory = Factory.class)
    public static final class FactoryAnnotatedValue {
        private final String value;

        FactoryAnnotatedValue(String value) {
            this.value = value;
        }

        @ToString
        public String asText() {
            return value;
        }
    }

    public static final class Factory {
        private Factory() {
        }

        @FromString
        public static FactoryAnnotatedValue parse(String value) {
            return new FactoryAnnotatedValue(value);
        }
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/AnnotationStringConverterFactoryTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import org.joda.convert.FromString
import org.joda.convert.FromStringFactory
import org.joda.convert.StringConvert
import org.joda.convert.ToString
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class AnnotationStringConverterFactoryTest {
  @Test
  def reachesAnnotatedConstructorInterfaceAndFactoryDiscoveryPaths(): Unit = {
    val convert: StringConvert = StringConvert.create()

    val stringConstructed: StringAnnotatedConstructorValue = convert.convertFromString(
      classOf[StringAnnotatedConstructorValue],
      "string-constructor"
    )
    val charSequenceConstructed: CharSequenceAnnotatedConstructorValue = convert.convertFromString(
      classOf[CharSequenceAnnotatedConstructorValue],
      "char-sequence-constructor"
    )
    val interfaceValue: InterfaceAnnotatedValue = new InterfaceAnnotatedValue("interface")
    val factoryFailure: IllegalStateException = assertThrows(
      classOf[IllegalStateException],
      () => convert.convertFromString(classOf[FactoryAnnotatedValue], "factory")
    )

    assertEquals("string-constructor", convert.convertToString(stringConstructed))
    assertEquals("char-sequence-constructor", convert.convertToString(charSequenceConstructed))
    assertEquals("interface", convert.convertToString(interfaceValue))
    assertEquals(
      s"Class annotated with @ToString but not with @FromString: ${classOf[FactoryAnnotatedValue].getName}",
      factoryFailure.getMessage
    )
  }
}

final class StringAnnotatedConstructorValue @FromString() (private val value: String) {
  @ToString
  def asText: String = value
}

final class CharSequenceAnnotatedConstructorValue @FromString() (private val value: CharSequence) {
  @ToString
  def asText: String = value.toString
}

trait ConvertibleAnnotatedInterface {
  @ToString
  def asText: String
}

final class InterfaceAnnotatedValue @FromString() (private val value: String)
    extends ConvertibleAnnotatedInterface {
  override def asText: String = value
}

@FromStringFactory(factory = classOf[FactoryAnnotatedValueFactory])
final class FactoryAnnotatedValue(private val value: String) {
  @ToString
  def asText: String = value
}

final class FactoryAnnotatedValueFactory private ()
